cmd/kubeadm/app/util/patches/patches.go

```diff
@@ -0,0 +1,338 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package patches
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	"sigs.k8s.io/yaml"
+)
+
+// PatchTarget defines a target to be patched, such as a control-plane static Pod.
+type PatchTarget struct {
+	// Name must be the name of a known target.
+	Name string
+
+	// StrategicMergePatchObject is only used for strategic merge patches.
+	// It represents the underlying object type that is patched - e.g. "v1.Pod"
+	StrategicMergePatchObject interface{}
+
+	// Data must contain the bytes that will be patched.
+	Data []byte
+}
+
+// PatchManager defines an object that can apply patches.
+type PatchManager struct {
+	patchSets    []*patchSet
+	knownTargets []string
+	output       io.Writer
+}
+
+// patchSet defines a set of patches of a certain type that can patch a PatchTarget.
+type patchSet struct {
+	targetName string
+	patchType  types.PatchType
+	patches    []string
+}
+
+// String() is used for unit-testing.
+func (ps *patchSet) String() string {
+	return fmt.Sprintf(
+		"{%q, %q, %#v}",
+		ps.targetName,
+		ps.patchType,
+		ps.patches,
+	)
+}
+
+var (
+	pathLock  = &sync.RWMutex{}
+	pathCache = map[string]*PatchManager{}
+
+	patchTypes = map[string]types.PatchType{
+		"json":      types.JSONPatchType,
+		"merge":     types.MergePatchType,
+		"strategic": types.StrategicMergePatchType,
+		"":          types.StrategicMergePatchType, // Treat an empty value as the default = strategic.
+	}
+)
+
+// GetPatchManagerForPath creates a patch manager that can be used to apply patches to "knownTargets".
+// "path" should contain patches that can be used to patch the "knownTargets".
+// If "output" is non-nil, messages about actions performed by the manager would go on this io.Writer.
+func GetPatchManagerForPath(path string, knownTargets []string, output io.Writer) (*PatchManager, error) {
+	pathLock.RLock()
```

Comment:
what about 
```suggestion
	pathLock.RLock()
	defer pathLock.RUnlock()
```
So we can assume we are in are in a locked context for the rest of this func?

Reply:
we have two parts of the logic here. one is lock / unlocking for read operations (near function start) and one is lock / unlocking for write operations (near function end).

`defer pathLock.RUnlock()` would not work because we need to "unlock for reading" as soon as the reading is done, so that we can "lock for writing".

in an ideal world, i would have made this API re-entrant leaving the management of "patches for a path" to the caller and not have cache at all. yet this is tricky given the kubeadm source is a bit of a mess WRT static Pod management and there is nowhere (e.g. no context) to store the underlying "patch manager".

Reply:
@fabriziopandini technically `defer` is what has to be used. However, it needs to be in a nested block.

Reply:
i'd argue the existing lock management is completely valid.

Reply:
Given that lock is ther for defensive programming but not for a concrete use case, handling read locks and write locks in separated code paths is kind of overengineering.
Said that I'm fine with current implementation; feel free to resolve the conversation